### PR TITLE
No call count reset

### PR DESCRIPTION
**Pull request by jonathan-benn**
Call count is not reset anymore on returns/throws

Fixes note35/sinon#32
Fixes note35/sinon#35

### CODE REVIEW

**Comment by note35** [comment]
I will review most of this PR after I merge previous one.

One question I can mention in advance is about lots similar code in `stub.py` and `Wrapper.py`.

For example: 
`__get_matching_indices` @ `stub.py` and `__get_call_indices` @ `Wrapper.py`
The difference of them is with or without conditions.

**Do you plan to refactor them in the future?**

I do believe it's necessary to do. But based on my memory, it may not be an easy because the passing arguments of `*args` and `**kwargs`.

**Comment by note35** [comment]
Hi jonathan,

I need your help for fixing the conflict of this PR!

--Kir

**Comment by jonathan-benn** [comment]
Hi Kir,

I thought I had fixed all the conflicts on Friday. I will investigate.

--Jonathan

**Comment by jonathan-benn** [comment]
Hi Kir,

I rebased on top of the current `note35:master` branch and that seems to have fixed the merge conflicts.

--Jonathan

**Comment by jonathan-benn** [comment]
Hi Kir, you asked:

> One question I can mention in advance is about lots similar code in `stub.py` and `Wrapper.py`.

I think perhaps this was an error in Github's algorithm for showing file changes. What happened is that I _moved_ the functions from `Wrapper.py` to `stub.py`. After rebasing, I think the change is more clear now in Github,

I think it makes more sense for the functions that calculate what return value the stub should return are in `stub.py` rather than the spy wrapper. If you disagree, I can move them back to `Wrapper.py`.

--Jonathan

**Comment by note35** [comment]
> I think it makes more sense for the functions that calculate what return value the stub should return are in stub.py rather than the spy wrapper. If you disagree, I can move them back to Wrapper.py.

Sorry for the delay review, I thought about this change for few days to figure out whether this change is proper or not, I can accept your offer. My previous design had a big problem, the `Wrapper` shouldn't contain **non-general function**.

`condition['action']` was apparently violating the philosophy.

File: sinon/lib/stub.py [hunk 1]
```diff
@@ -40,20 +180,18 @@ def _append_condition(self, sinon_stub_condition, func):
...
             sinon_stub_condition: the _SinonStubCondition object that holds the current conditions
-            func: returns a value or raises an exception, as specified by the user
-        Returns: the SinonStub._conditions dictionary (for convenience)
+            func: returns a value or raises an exception (i.e. the action to take, as specified by the user)
         '''
         self._conditions["args"].append(sinon_stub_condition._cond_args)
         self._conditions["kwargs"].append(sinon_stub_condition._cond_kwargs)
         self._conditions["oncall"].append(sinon_stub_condition._oncall)
-        self._conditions["action"].append(func)
-        return self._conditions
```
Comment (note35) [inline sinon/lib/stub.py:56]: nice catch on another bad design!

**Review by note35 — APPROVED** [review]
Thanks for your efforts again! This change is perfect without any issue!

**Pull request merged**